Ensure custody of timezones from query to results, add debugging info

senior opened the pull request:
This PR makes several changes intended to fix timezone related
issues and expose more information when there are problems. For
databases that support report timezones, this commit fixes an issue
where the JVM's timezone was used to interpret datetimes that didn't
include a timezone. The report timezone should be used instead. This
commit also adds a step to the sync process to detect and store the
databases timezone, which will be helpful in troubleshooting timezone
issues with a new timezone debugging page (to be added later).

iethree committed:
Ensure custody of timezones from query to results, add debugging info

This commit makes several changes intended to fix timezone related
issues and expose more information when there are problems. For
databases that support report timezones, this commit fixes an issue
where the JVM's timezone was used to interpret datetimes that didn't
include a timezone. The report timezone should be used instead. This
commit also adds a step to the sync process to detect and store the
databases timezone, which will be helpful in troubleshooting timezone
issues with a new timezone debugging page (to be added later).

salsakran commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -383,31 +420,8 @@
-(defn- run-query-without-timezone [driver settings connection query]
-  (do-in-transaction connection (partial run-query query)))
-
-(defn- run-query-with-timezone [driver settings connection query]
```
Dumb question, but does this affect date extracts (which should be done with respect to the reporting timezone)?

iethree committed:
Add unit test workarounds for oracle report timezone

The issue is the date/timestamp functions we use in Oracle don't apply
the report timezone that we have set in the Oracle session. This
commit special cases those oracle date test results until we can get
the issue fixed, written up as #5729.

iethree committed:
Fix report-timezones and Presto [ci drivers]

iethree committed:
Minor fixes for reflection and a long line in the mysql tests

camsaul commented on test/metabase/sync/sync_metadata/sync_timezone_test.clj:
```diff
@@ -0,0 +1,27 @@
+             [metabase.sync.sync-metadata.sync-timezone :as sync-tz]
+             [metabase.test.data.datasets :as datasets]
+             [metabase.test.data :as data]
+             [clj-time.core :as time]
```
ns sorting 😉

camsaul commented on test/metabase/sync/sync_metadata/sync_timezone_test.clj:
```diff
@@ -0,0 +1,27 @@
+(defn- db-timezone [db-or-id]
+  (db/select-one-field :timezone Database :id (u/get-id db-or-id)))
+
+(datasets/expect-with-engines #{:h2 :postgres}
```
what exactly are we testing here? A quick comment would be nice since otherwise it's not really clear what's going on

salsakran commented on src/metabase/driver.clj:
```diff
@@ -203,7 +227,8 @@
                                          (NoSuchMethodException.
-                                          (str (name driver) " does not implement table-rows-seq."))))})
+                                          (str (name driver) " does not implement table-rows-seq."))))
+   :current-db-time                   (constantly nil)})
```
is this time or timezone?

camsaul commented on test/metabase/query_processor_test/date_bucketing_test.clj:
```diff
@@ -226,27 +294,41 @@
+    [[23 49] [24 47] [25 39] [26 58] [27 7]]
+
+
+
```
why so many blank lines?

salsakran commented on src/metabase/driver/presto.clj:
```diff
@@ -40,25 +42,34 @@
 
 (defn- parse-time-with-tz [s]
   ;; Try parsing with offset first then with full ZoneId
+  (println "getting ready to parse " s)
```
I think we'd want to remove this before merging.

senior replied:
Could have swore I got rid of those. Good catch, I'll remove them

camsaul replied:
assuming this is left over from debugging? You can leave it there just make it `log/debug` or something

salsakran commented on src/metabase/driver/presto.clj:
```diff
@@ -40,25 +42,34 @@
 
 (defn- parse-timestamp-with-tz [s]
   ;; Try parsing with offset first then with full ZoneId
+  (println "getting ready to parse " s)
```
I think we'd want to remove this before merging.

camsaul replied:
same thing, `log/debug`

camsaul commented on test/metabase/driver/postgres_test.clj:
```diff
@@ -7,9 +7,7 @@
              [util :as u]]
-            [metabase.driver
-             [generic-sql :as sql]
-             postgres]
```
`cljr-clean-namespace` probably did this for you automatically but the reason we're loading the postgres driver here is that the `import` statement below it will fail if some other namespace hasn't already loaded the postgres driver. Thus something like `lein test metabase.driver.postgres-test` could potentially fail.

It's unfortunate that there's no way to tell `cljr-clean-ns` that some of those lines are there on purpose like that. But I would just keep that in mind when you review your own PR, if there's a required namespace that doesn't look like it's being used for something being removed take a look at the `:import`s as well

senior replied:
Yeah, I've been having issues with that, running cljr-clean-ns, then it breaking the code. I'll add this back.

camsaul replied:
Yeah I would just try to review all the changes in your code when you're wrapping up a PR and look for things like that and fix as needed. `cljr-clean-ns` is still better than doing things yourself since now it's a manual review of an automatic process instead of a manual review of a manual process

camsaul commented on src/metabase/util.clj:
```diff
@@ -858,3 +858,37 @@
+  attempts fail before the right one is found. Using this retains that
+  flexibility but improves performance by trying the most likely ones
+  first"
+  (let [most-likely-default-formatters [:mysql :date-hour-minute-second :date-time :date
```
👍

salsakran reviewed (approved):
Minor comments, but I would really like a little more description of what is being tested. Just changing the bin counts makes it hard to follow. 

It would be especially useful to somehow explicitly call out reporting timezone, database timezone, and jvm timezone interactions in the the test suite.

camsaul commented on src/metabase/sync/fetch_metadata.clj:
```diff
@@ -22,3 +23,8 @@
     (when (driver/driver-supports? driver :foreign-keys)
       (driver/describe-table-fks driver database table))))
+
+(s/defn ^:always-validate db-timezone :- s/Str
```
It might be possible to define an output schema a little stricter than just `s/Str` here. You could do `su/NonBlankString` or even better define a new schema that checks whether the timezone is legit. You could do something like check whether `(java.util.TimeZone/getTimeZone <str>)` works correctly (for an invalid timezone it just returns `GMT` so you could check whether the `TimeZone` you get back isn't GMT, unless of course the string passed in was `"GMT"`, in which case that would be ok)

camsaul commented on src/metabase/driver/presto.clj:
```diff
@@ -18,7 +18,9 @@
-             [ssh :as ssh]])
+             [ssh :as ssh]]
+            [clj-time.format :as tformat]
+            [clj-time.core :as time])
```
ns sorting

iethree committed:
Minor changes for code cleanup

senior commented:
@salsakran A description would have been useful for me too as I was making these changes. I'll do some research and write up some comments around the test.

As far as assertions around timezones in the test suite, I've added a test for each database asserting the timezone that the database is expected to be in for the test. With the date tests (date bucketing and unix timestamp) I've added an explicit reporting timezone (pacific). I could add an assertion for the JVM timezone to the date bucketing tests to cover that. Is that the kind of thing you were looking for?

salsakran commented:
I was thinking along the lines of:
if the jvm timezone is Central, the db timezone is pacific and the reporting timezone is eastern, for these 3 rows grouped by day, are all the day counts correct.

iethree committed:
Add time zone string schema validation

iethree committed:
Increase the docs and timezone related date bucketing tests

This commit recasts many of the existing date bucketing tests into
several different reporting and JVM timezones with expectations
adjusted to match. This commit also adds quite a few docs around the
various tests with explainations of what it's testing, why timezones
impact it and even some tests to help with debugging these timezone
related issues.

iethree committed:
Merge remote-tracking branch 'upstream/master' into sync-db-timezone [ci drivers]

iethree committed:
Add exceptions for SQLServer when changing the JVM timezone

The new SQLServer driver introduces some new timezone behavior. If the
JVM is in a different timezone than the SQL Server instance, you can
get incorrect behavior as the timestamps are assumed to be in the JVM
timezone. Added some exceptions to the test cases to avoid that situation.

iethree committed:
Minor fixups to the new date bucketing tests [ci drivers]

iethree committed:
Add redshift to the date-bucket-test exceptions list [ci drivers]

Redshift doesn't handle changing the JVMs timezone to something
different than what the database's timezone is (similar to
SQLServer). This adds that to the list so we don't try and run the JVM
timezone changing tests against it.

iethree committed:
Added similar date bucketing test exceptions for JVM timezone [ci drivers]

Similar to the exceptions needed for SQLServer and H2 when changing
JVM timezones, Mongo is also sensitive to this, so the tests should be
run on Mongo.

iethree committed:
Merge remote-tracking branch 'upstream/master' into sync-db-timezone

camsaul reviewed (approved):
LGTM

mazameli commented:
@senior, does this also resolve:
#3642 
#3744 
#4180
#4245
#2517
#4372
and #2035?

Those issues were all noted as being dupes according to @salsakran, and 2035 as being the "root cause." Should these be closed as well?

senior commented:
#3642 - Looking at @tlrobinson's comment (here)[https://github.com/metabase/metabase/issues/3642#issuecomment-265856616] it looks like it's an issue on the UI side, query results that contain multiple offsets, such as a daylight savings shift in the middle of a resultset

#3744 - Fixed with this PR

#4180 not sure, my guess would be not as the druid stuff goes through a very different code path, worth testing though

#4245 - I'll validate this, it looks like the correct data is coming back, but being displayed incorrectly

#2517 - Fixed with this PR

#4372 - No but I put some advice on addressing this through what we have today

#2035 - Fixed with this PR
